copystructure.go

```diff
@@ -48,14 +49,25 @@ type walker struct {
 	vals        []reflect.Value
 	cs          []reflect.Value
 	ps          []bool
+	locks       []sync.Locker
 }
 
 func (w *walker) Enter(l reflectwalk.Location) error {
 	w.depth++
+	// this normally appends only once, but make sure we start with enough
+	// space to use depth as an index
+	for w.depth >= len(w.locks) {
+		w.locks = append(w.locks, nil)
+	}
```

Comment:
You can use `cap` here to do this w/o the `append` magic since the zero value of this slice is going to be `nil` anyways.